rcl/src/rcl/time.c

```diff
@@ -453,19 +454,18 @@ rcl_clock_remove_jump_callback(
   }
 
   // Shrink size of the callback array
-  if (clock->num_jump_callbacks == 1) {
+  if (clock->num_jump_callbacks-- == 1) {
     clock->allocator.deallocate(clock->jump_callbacks, clock->allocator.state);
     clock->jump_callbacks = NULL;
   } else {
     rcl_jump_callback_info_t * callbacks = clock->allocator.reallocate(
-      clock->jump_callbacks, sizeof(rcl_jump_callback_info_t) * (clock->num_jump_callbacks - 1),
+      clock->jump_callbacks, sizeof(rcl_jump_callback_info_t) * clock->num_jump_callbacks,
```

Comment:
I'm having a hard time understanding what is happening here. At this point `clock->num_jump_callbacks` is 0. So realloc(ptr, 0) is happening. Why not just free the array? 

It looks like realloc of size 0 is now undefined behavior. I think this `if` statement could be cleaned up.

http://www.open-std.org/jtc1/sc22/wg14/www/docs/summary.htm#dr_400

Reply:
By undefined, I guess it's implementation defined. So this might not be portable.

Reply:
> At this point `clock->num_jump_callbacks` is `0`

Hmm for `clock->num_jump_callbacks` to be `0` here, it would've had to be `1` when the condition was evaluated, in which case it would've executed the first branch. Note that we're using postfix decrement, not prefix. I can change the condition if it's confusing though.

Reply:
You're right, I was confused when looking at the diff of this logic. The if clause is executed if num_jump_callbacks is 1 when entering the loop. It will never be 0 because this function will have returned before then. The else clause is evaluated when it is strictly greater than 1. Thus reallocate will be called with a size that is never 0.

The only other difference that might be important between the diffs is that if reallocate fails, `num_jump_callbacks` no longer matches the size of `jump_callbacks`. Should num_jump_callbacks be reincremented in that case?

Reply:
> Should `num_jump_callbacks` be reincremented in that case?

I thought about it, but the [preceding loop](https://github.com/ros2/rcl/blob/7723308e7c7f7515590bd709525fd840b4aed254/rcl/src/rcl/time.c#L445) probably removed that callback by then. 

Also, a failed reallocation means the function failed to shrink storage, not that the callback itself wasn't removed successfully. A future deallocation wouldn't be affected and a future reallocation would be a no-op.